pkg/server/backuprestoreserver.go

```diff
@@ -128,67 +138,105 @@ func (b *BackupRestoreServer) startHTTPServer(initializer initializer.Initialize
+					return
+				}
 
-	etcdInitializer := initializer.NewInitializer(restoreOpts, b.config.SnapstoreConfig, b.logger.Logger)
+				// set "http handler" with the latest snapshotter object
+				handler.SetSnapshotter(ssr)
+				defragCallBack = ssr.TriggerFullSnapshot
+				go handleSsrStopRequest(leCtx, handler, ssr, ackCh, ssrStopCh)
+			}
+			go b.runEtcdProbeLoopWithSnapshotter(leCtx, handler, ssr, ss, ssrStopCh, ackCh)
+			go defragmentor.DefragDataPeriodically(leCtx, b.config.EtcdConnectionConfig, b.defragmentationSchedule, defragCallBack, b.logger)
+		},
+		OnStoppedLeading: func() {
+			// stops the running snapshotter
+			if runServerWithSnapshotter {
+				ssr.SsrStateMutex.Lock()
+				defer ssr.SsrStateMutex.Unlock()
+				if ssr.SsrState == brtypes.SnapshotterActive {
+					ssrStopCh <- emptyStruct
+					b.logger.Info("backup-restore stops leading...")
+				}
+				handler.SetSnapshotterToNil()
+			}
+		},
+	}
 
-	b.logger.Infof("Creating snapstore from provider: %s", b.config.SnapstoreConfig.Provider)
-	ss, err := snapstore.GetSnapstore(b.config.SnapstoreConfig)
-	if err != nil {
-		return fmt.Errorf("failed to create snapstore from configured storage provider: %v", err)
+	memberLeaseCallbacks := &leaderelection.MemberLeaseCallbacks{
+		StartLeaseRenewal: func() {
+			mmStopCh = make(chan struct{})
+			if b.config.HealthConfig.MemberLeaseRenewalEnabled {
+				go heartbeat.RenewMemberLeasePeriodically(ctx, mmStopCh, b.config.HealthConfig, b.logger, b.config.EtcdConnectionConfig)
+			}
+		},
+		StopLeaseRenewal: func() {
+			if b.config.HealthConfig.MemberLeaseRenewalEnabled {
+				mmStopCh <- emptyStruct
+			}
+		},
 	}
 
-	b.logger.Infof("Creating snapshotter...")
-	ssr, err := snapshotter.NewSnapshotter(b.logger, b.config.SnapshotterConfig, ss, b.config.EtcdConnectionConfig, b.config.CompressionConfig, b.config.HealthConfig)
+	checkLeadershipFunc := leaderelection.IsLeader
+
+	b.logger.Infof("Creating leaderElector...")
+	le, err := leaderelection.NewLeaderElector(b.logger, b.config.EtcdConnectionConfig, b.config.LeaderElectionConfig, leaderCallbacks, memberLeaseCallbacks, checkLeadershipFunc)
 	if err != nil {
 		return err
 	}
 
-	handler := b.startHTTPServer(etcdInitializer, ssr)
-	defer handler.Stop()
-
-	ssrStopCh := make(chan struct{})
-	go handleSsrStopRequest(ctx, handler, ssr, ackCh, ssrStopCh)
-	go handleAckState(handler, ackCh)
-
-	go defragmentor.DefragDataPeriodically(ctx, b.config.EtcdConnectionConfig, b.defragmentationSchedule, ssr.TriggerFullSnapshot, b.logger)
+	if runServerWithSnapshotter {
+		go handleAckState(handler, ackCh)
+	}
 
 	//TODO @aaronfern: Add functionality for member garbage collection
 	if b.config.HealthConfig.MemberLeaseRenewalEnabled {
-		go heartbeat.RenewMemberLeasePeriodically(ctx, b.config.HealthConfig, b.logger, b.config.EtcdConnectionConfig)
+		go heartbeat.RenewMemberLeasePeriodically(ctx, mmStopCh, b.config.HealthConfig, b.logger, b.config.EtcdConnectionConfig)
 	}
```

Comment:
Why is this required if `memberLeaseCallbacks` already does the same task?

Reply:
Actually all backup-restore have default state as `Follower`, hence `RenewMemberLeasePeriodically()` starts for everybody  and `memberLeaseCallbacks` is going to take care when there is any state change(refer [doc](https://github.com/gardener/etcd-backup-restore/blob/eb57e66f058b3176c2720a5504e84e91fb1e2cf5/doc/concepts/leader_election.md#complete-work-flow-leader-election-state-diagram)) in backup-restore.

Reply:
Got it. Thanks for explaining 👍🏻